src/builtins/BuiltinShadowRealm.cpp

```diff
@@ -0,0 +1,107 @@
+#include "Escargot.h"
+#include "interpreter/ByteCode.h"
+#include "interpreter/ByteCodeGenerator.h"
+#include "interpreter/ByteCodeInterpreter.h"
+#include "runtime/Global.h"
+#include "runtime/GlobalObject.h"
+#include "runtime/Context.h"
+#include "runtime/Environment.h"
+#include "runtime/EnvironmentRecord.h"
+#include "runtime/ExtendedNativeFunctionObject.h"
+#include "runtime/VMInstance.h"
+#include "runtime/ShadowRealmObject.h"
+#include "runtime/NativeFunctionObject.h"
+#include "runtime/ScriptFunctionObject.h"
+#include "parser/Script.h"
+#include "parser/ScriptParser.h"
+
+namespace Escargot {
+
+#if defined(ESCARGOT_ENABLE_SHADOWREALM)
+
+//  https://tc39.es/proposal-shadowrealm/#sec-shadowrealm
+static Value builtinShadowRealmConstructor(ExecutionState& state, Value thisValue, size_t argc, Value* argv, Optional<Object*> newTarget)
+{
+    // If NewTarget is undefined, throw a TypeError exception.
+    if (!newTarget) {
+        ErrorObject::throwBuiltinError(state, ErrorCode::TypeError, ErrorObject::Messages::GlobalObject_ConstructorRequiresNew);
+    }
+
+    // Let O be ? OrdinaryCreateFromConstructor(NewTarget, "%ShadowRealm.prototype%", « [[ShadowRealm]] »).
+    Object* proto = Object::getPrototypeFromConstructor(state, newTarget.value(), [](ExecutionState& state, Context* constructorRealm) -> Object* {
+        return constructorRealm->globalObject()->shadowRealmPrototype();
+    });
+    ShadowRealmObject* O = new ShadowRealmObject(state, proto);
+
+    // Let callerContext be the running execution context.
+    // Perform ? InitializeHostDefinedRealm().
+    // Let innerContext be the running execution context.
+    Context* innerContext = new Context(state.context()->vmInstance());
+    // Remove innerContext from the execution context stack and restore callerContext as the running execution context.
+    // Let realmRec be the Realm of innerContext.
+    // Set O.[[ShadowRealm]] to realmRec.
+    O->setRealmContext(innerContext);
+    // Perform ? HostInitializeShadowRealm(realmRec, innerContext, O).
+    // Assert: realmRec.[[GlobalObject]] is an ordinary object.
+    ASSERT(innerContext->globalObject()->isOrdinary());
+    // Return O.
+    return O;
+}
+
+static Value builtinShadowRealmEvaluate(ExecutionState& state, Value thisValue, size_t argc, Value* argv, Optional<Object*> newTarget)
+{
+    // Let O be the this value.
+    const Value& O = thisValue;
+    // Perform ? ValidateShadowRealmObject(O).
+    if (!O.asObject()->isShadowRealmObject()) {
```

Comment:
Since `O` is a Value, it could be a non-object value.
If `O` is not an object, then `asObject()` would trigger an assertion error.
`isObject()` should be invoked first for this case